File: viewer/src/model/metaRules.ts
/**
 * タグの正規化と検証（P-12）。サーバー側の erd.core.model.MetaRules と同じ規則を持つ。
 * 両側に置くのは、UI で防ぎつつ、手で編集したファイルや古いクライアントの書き込みも
 * サーバーで弾くため（片側だけだと必ず壊れた値が入る）。
 */

/** 1タグの最大長（コードポイント数）。 */
export const MAX_TAG_LENGTH = 32;

/** 1テーブル / 1カラムに付けられるタグの最大数。 */
export const MAX_TAGS = 20;

/** タグの区切りとして扱う文字（空白でタグを確定する UI に合わせる）。 */
const SEPARATORS = /[\s,、]+/u;

/** NFC 正規化 + 前後空白の除去。 */
export function normalizeTag(raw: string): string {
  return raw.normalize("NFC").trim();
}

/** 正規化 → 空要素の除去 → 重複除去（大小無視・先勝ち）。並びは入力順を保つ。 */
export function normalizeTags(raw: readonly string[]): string[] {
  const out: string[] = [];
  const seen = new Set<string>();
  for (const tag of raw) {
    const t = normalizeTag(tag);
    if (t === "") continue;
    const key = t.toLowerCase();
    if (seen.has(key)) continue;
    seen.add(key);
    out.push(t);
  }
  return out;
}

/**
 * 貼り付け・一括入力の分割（`core, auth 廃止` → 3件）。
 * 確定キー（空白）と同じ区切りで割るので、貼り付けと手入力の結果が一致する。
 */
export function splitTagInput(text: string): string[] {
  return normalizeTags(text.split(SEPARATORS));
}

export type TagErrorCode = "TAG_TOO_LONG" | "TAG_DUPLICATE";

/**
 * 確定しようとしているタグ1件の検証。問題なければ null。
 * 空白・区切り文字は {@link splitTagInput} で分割済みのため、ここでは長さと重複だけを見る。
 */
export function tagError(tag: string, existing: readonly string[]): TagErrorCode | null {
  if ([...tag].length > MAX_TAG_LENGTH) return "TAG_TOO_LONG";
  const key = tag.toLowerCase();
  if (existing.some((t) => t.toLowerCase() === key)) return "TAG_DUPLICATE";
  return null;
}

File: server/src/main/java/erd/core/model/MetaRules.java
package erd.core.model;

import java.text.Normalizer;
import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;

/**
 * タグ・色の正規化と検証の規則（P-12 / P-13）。ビューアにも同じ規則の実装がある
 * （viewer: model/metaRules.ts）。<b>サーバー側にも置くのは、手で編集したファイルや
 * 古いクライアントからの書き込みで壊れた値が入るのを防ぐため</b>。
 *
 * <p>正規化（黙って直すもの）と検証（422 で拒否するもの）を分けている。前後の空白や
 * 重複は書き手の意図が明らかなので直し、区切り文字の混入や長すぎる値は
 * 「入力の取り違え」の可能性が高いので拒否する。
 */
public final class MetaRules {

    /** 1タグの最大長（文字数。コードポイント数で数える）。 */
    public static final int MAX_TAG_LENGTH = 32;

    /** 1テーブル / 1カラムに付けられるタグの最大数。 */
    public static final int MAX_TAGS = 20;

    private MetaRules() {
    }

    /**
     * タグ配列の正規化: NFC → 前後空白の除去 → 空要素の除去 → 重複除去（大小無視・先勝ち）。
     * 並びは入力順を保つ（タグの順序に意味は無いが、人の書いた順を勝手に変えない）。
     */
    public static List<String> normalizeTags(List<String> raw) {
        if (raw == null || raw.isEmpty()) return List.of();
        List<String> out = new ArrayList<>();
        Set<String> seen = new LinkedHashSet<>();
        for (String tag : raw) {
            if (tag == null) continue;
            String t = Normalizer.normalize(tag, Normalizer.Form.NFC).trim();
            if (t.isEmpty()) continue;
            if (seen.add(t.toLowerCase(Locale.ROOT))) out.add(t);
        }
        return List.copyOf(out);
    }

    /** 色の正規化: 前後空白の除去。空文字は「未設定」= null に倒す。 */
    public static String normalizeColor(String raw) {
        if (raw == null) return null;
        String c = raw.trim();
        return c.isEmpty() ? null : c;
    }

    /**
     * 正規化済みタグ1件の検証。問題なければ null、あればエラーコードを返す。
     *
     * <p>空白を禁じるのは、タグ入力 UI が空白でタグを確定するため（P-12）。
     * 区切り文字（{@code ,} {@code 、}）を禁じるのは、1件のつもりで複数件を書いた
     * 入力を黙って1件のタグにしないため。
     */
    public static String tagError(String tag) {
        if (tag.codePoints().anyMatch(Character::isWhitespace)) return "TAG_WHITESPACE";
        if (tag.indexOf(',') >= 0 || tag.indexOf('、') >= 0) return "TAG_SEPARATOR";
        if (tag.codePointCount(0, tag.length()) > MAX_TAG_LENGTH) return "TAG_TOO_LONG";
        return null;
    }
}
